server/dataset/cxg_dataset.py

```diff
@@ -612,46 +1036,193 @@ def _get_schema(self):
+            geneset_entries, tid = raw[0], int(raw[1])
+        else:
+            geneset_entries = raw
+
+        normalized_genesets = []
+        if isinstance(geneset_entries, list):
+            for item in geneset_entries:
+                if isinstance(item, dict):
+                    normalized_genesets.append(item)
+                elif isinstance(item, (list, tuple)) and len(item) == 2 and isinstance(item[1], dict):
+                    geneset_name = item[0]
+                    geneset_obj = item[1]
+                    if "geneset_name" not in geneset_obj:
+                        geneset_obj = {**geneset_obj, "geneset_name": geneset_name}
+                    normalized_genesets.append(geneset_obj)
+        elif isinstance(geneset_entries, dict):
+            for name, value in geneset_entries.items():
+                if isinstance(value, dict):
+                    if "geneset_name" not in value:
+                        value = {**value, "geneset_name": name}
+                    normalized_genesets.append(value)
+
+        return {"genesets": normalized_genesets, "tid": tid}
+
+    def get_genesets(
+        self,
+        user_id: Optional[str] = None,
+    ):
+        saved = self.get_saved_gene_sets(user_id=user_id)
+        if saved is not None:
+            return saved
 
-    def get_genesets(self):
         if self.genesets is None:
             with self.lock:
                 self.genesets = self._get_genesets()
-        return self.genesets
-
-    def annotation_to_fbs_matrix(self, axis, fields=None, num_bins=None):
+        if isinstance(self.genesets, dict):
+            geneset_entries = self.genesets.get("genesets", [])
+            normalized_genesets = []
+            if isinstance(geneset_entries, list):
+                for item in geneset_entries:
+                    if isinstance(item, dict):
+                        normalized_genesets.append(item)
+                    elif isinstance(item, (list, tuple)) and len(item) == 2 and isinstance(item[1], dict):
+                        geneset_name = item[0]
+                        geneset_obj = item[1]
+                        if "geneset_name" not in geneset_obj:
+                            geneset_obj = {**geneset_obj, "geneset_name": geneset_name}
+                        normalized_genesets.append(geneset_obj)
+            elif isinstance(geneset_entries, dict):
+                for name, value in geneset_entries.items():
+                    if isinstance(value, dict):
+                        if "geneset_name" not in value:
+                            value = {**value, "geneset_name": name}
+                        normalized_genesets.append(value)
+            return {
+                "genesets": normalized_genesets,
+                "tid": self.genesets.get("tid", 0),
+            }
+        # Fallback - should not happen due to normalization above.
+        return {"genesets": [], "tid": 0}
+
+    def annotation_to_fbs_matrix(
```

Comment:
Merges persisted user annotations over base obs (fields subset supported). Please review performance for large obs and correctness when filtering fields.